### DISCUSSION TITLE
People counter with 2 VL53L0X modules - Help converting yaml to Tasmota script

### DISCUSSION THREAD

**Discussion by orgnetic** [arendst/Tasmota#18082] [Q&A]
Hello everyone!

I'm a newbie on Tasmota scritping language, and I don't have a clue on how to convert the following ESPhome code. 

This corresponds to a tried and tested solution, that can be of use to many Tasmota lovers.

If anyone can help me, I would be glad to offer a beer when visiting Portugal :)

So, without further ado, here is the code:

globals:
  - id: seq1
    type: int
    restore_value: no
    initial_value: '0'
  - id: seq2
    type: int
    restore_value: no
    initial_value: '0'
  - id: seq3
    type: int
    restore_value: no
    initial_value: '0'
  - id: seq4
    type: int
    restore_value: no
    initial_value: '0'
    
  - id: count
    type: int
    restore_value: no
    initial_value: '0'
    
  - id: inside_previous_value
    type: bool
    restore_value: no
    initial_value: 'false'
    
  - id: outside_previous_value
    type: bool
    restore_value: no
    initial_value: "false"

binary_sensor:
  - platform: template
    id: lab_inside_door_laser
    name: "Lab Inside Door Laser"
    lambda: !lambda |-
        if((id(sens_1).state < 0.7f) != id(inside_previous_value)){
          id(inside_previous_value) = (id(sens_1).state < 0.7f);
          if(((id(sens_1).state < 0.7f) == false) && ((id(sens_2).state < 0.7f) == false)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 0;
          }else if(((id(sens_1).state < 0.7f) == false) && ((id(sens_2).state < 0.7f) == true)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 1;
          }else if(((id(sens_1).state < 0.7f) == true) && ((id(sens_2).state < 0.7f) == false)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 2;
          }else if(((id(sens_1).state < 0.7f) == true) && ((id(sens_2).state < 0.7f) == true)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 3;
          }
          if((id(seq1) == 1) && (id(seq2) == 3) && (id(seq3) == 2) && (id(seq4) == 0)){
            id(count) ++;
            id(people_in_lab).publish_state(id(count));
          }else if((id(seq1) == 2) && (id(seq2) == 3) && (id(seq3) == 1) && (id(seq4) == 0)){
            id(count) --;
            if(id(count) < 0){
              id(count) = 0;
            }
            id(people_in_lab).publish_state(id(count));
          }
        }
        return id(sens_1).state < 0.7f;
        
        

  - platform: template
    id: lab_outside_door_laser
    name: "Lab Outside Door Laser"
    lambda: !lambda |-
        if((id(sens_2).state < 0.7f) != id(outside_previous_value)){
          id(outside_previous_value) = (id(sens_2).state < 0.7f);
          if(((id(sens_1).state < 0.7f) == false) && ((id(sens_2).state < 0.7f) == false)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 0;
          }else if(((id(sens_1).state < 0.7f) == false) && ((id(sens_2).state < 0.7f) == true)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 1;
          }else if(((id(sens_1).state < 0.7f) == true) && ((id(sens_2).state < 0.7f) == false)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 2;
          }else if(((id(sens_1).state < 0.7f) == true) && ((id(sens_2).state < 0.7f) == true)){
            id(seq1) = id(seq2);
            id(seq2) = id(seq3);
            id(seq3) = id(seq4);
            id(seq4) = 3;
          }
          if((id(seq1) == 1) && (id(seq2) == 3) && (id(seq3) == 2) && (id(seq4) == 0)){
            id(count)++;
            id(people_in_lab).publish_state(id(count));
          }else if((id(seq1) == 2) && (id(seq2) == 3) && (id(seq3) == 1) && (id(seq4) == 0)){
            id(count)--;
            if(id(count) < 0){
              id(count) = 0;
            }
            id(people_in_lab).publish_state(id(count));
          }
        }
        return id(sens_2).state < 0.7f;
        
        
  - platform: gpio
    pin: GPIO2
    name: "Lab PIR Sensor"
    device_class: motion

sensor:
  - platform: template
    id: people_in_lab
    name: "People in Lab"
    lambda: !lambda |-
      return id(count);
        

  - platform: vl53l0x
    name: "Sensor 1"
    id: sens_1
    i2c_id: bus_a
    address: 0x41
    update_interval: 50ms
    enable_pin: GPIO25
    timeout: 500us
    internal: true
    
    
  - platform: vl53l0x
    name: "Sensor 2"
    id: sens_2
    i2c_id: bus_a
    address: 0x42
    update_interval: 50ms
    enable_pin: GPIO32
    timeout: 500us
    internal: true`

The code is not mine, and for the sake of the authors recognition, here is the discussion:

https://community.home-assistant.io/t/people-counter-passing-through-a-door-using-vl53l0x-or-vl53l1x/312897

Cheers !

**Comment by barbudor** [comment]
Obviously this is using a state machine which would be difficult to reproduce in Tasmota without using [Scripting-Language](https://tasmota.github.io/docs/Scripting-Language/) or [Berry](https://tasmota.github.io/docs/Berry/)

Then by default, the Tasmota VL53L0X driver is performing a median filter with a length of 5 and a measurment every 250 ms
I'm not sure this would be fast enough 
If you want to remove that you need to recompile yourself and comment line 66:
https://github.com/arendst/Tasmota/blob/83f039cdf796358d3d9b5f225c4015089b0167bb/tasmota/tasmota_xsns_sensor/xsns_45_vl53l0x.ino#L66

What is the distance between the 2 sensors ?
How short do you estimate the time between somone crossing the 1st sensor then the 2nd sensor ?

**Comment by orgnetic** [comment]
Well, the distance between each other is 60mm, from the center of each sensor.

By default, the update interval of each sensor is 50ms, although some people use 10ms,

Regarding time, to detect a fast person, we would count on a walking pace of 5km/hour.

So, for a realiable solution with fast persons (or running kids), we're talking of ~1.4 meters per second, and that means we must count on ~43ms, between each person crossing.

PS: If it helps, there's a solution, based on one VL53L1X module (this one has the possibility to measure on separate zones), that can be seen here: 

[https://github.com/Lyr3x/Roode](https://github.com/Lyr3x/Roode)

Is this information useful?

**Reply by sfromis** [reply]
The way Tasmota handles sensors is primarily designed for picking up values like illuminance or temperature, where it is just fine to read values at intervals of about one second. Fast polling of I2C sensors at millisecond level is outside of what standard I2C drivers facilitate, and the current VL53L0X driver does not support using the IRQ pins for signals, either.

Thus, "further work" would be needed for reasonably accurate people-counting based on VL53L0X (or similar).

**Comment by orgnetic** [comment]
Well, I have compiled the driver with the #define VL53L0X_HIGH_SPEED option, The official documententation states that HIGH_SPEED mode has an update of around 20 ms.
Although, I understand that this doesn't mean that Tasmota will follow this speed.
Another approach would be to enable a bigger spacing between the 2 sensors, in a way that we could reach at least 100ms.

Would this be possible?

**Reply by sfromis** [reply]
Answers to that would be based on experimentation.

As I understand `VL53L0X_HIGH_SPEED`, it will allow the sensor to react faster (20 ms instead of 33 ms, but less accurate) when Tasmota reads a value; how often Tasmota will perform a read is another question.

**Comment by barbudor** [comment]
In Tasmota, binary imputs like switches/buttons are handled faster
So it could be more easy to achieve that using a focus IR beam to a pair of IR receiver to trigger a GPIO configured as switches
I would still be concerned by the speed of reaction required for that which could probabably not be achievable in Rules.

**Reply by sfromis** [reply]
With a digital input, you can also config it as a Counter pin, which can count pulses pretty quickly. This is the fastest way if you do not need to do processing of every pulse.

**Reply by barbudor** [reply]
Or more likely expect the 2 signals to behave like the quadrature signals of a rotary encoder to count up and down

**Comment by orgnetic** [comment]
First of all, I would like to emphasize how happy I am to see you guys trying to help me. Thank you very much for sharing your thoughts on this subject.

Reliable people counters are expensive devices, and most of them counts everything that passes the door.

The reason why I would like to base it on VL53LxX modules, is that by taking measures, we can roughly make a filter for children, pets and adults.

And obviously, I'm not willing to move to ESPhome, since Tasmota meets 99% of my expectations as a stand-alone, programmable smart device, that behave and works perfectly, without any connection to Home Assistant, for example.

Anyway, as a last resort, two of those [photoelectric switches](https://wiki.dfrobot.com/Gravity_Digital_Adjustable_Infrared_Proximity_Sensor_SKU_SEN0239) can be used. I just don't have any ideia on how to implement/wire them, as a quadrature rotary encoder.

@sfromis refers this code, on another subjet [https://github.com/arendst/Tasmota/discussions/17673#discussioncomment-5019729](https://github.com/arendst/Tasmota/discussions/17673#discussioncomment-5019729)

`on Rotary1#Pos1=1 do Add1 1 endon
on Rotary1#Pos1=0 do Sub1 1 endon`

Is this the right way to implement a people counter using Tasmota?
If so, it's just a matter of wiring the photoelectric switches just like a rotary encoder, right?

**Reply by sfromis** [reply]
The right way is what works :-) 

The rule I suggested was to count up or down for twisting a "real" rotary encoder. You could experiment with tickling two digital inputs by other means, like buttons. Or by generating pulses on "relay" outputs connected to the inputs. Or whatever.

**Comment by ozett** [comment]
in berry-script on an ESP32 you could try to use skeleton-code for a driver and repeat in 50ms loop. see this timings:
https://tasmota.github.io/docs/Berry/#creating-a-tasmota-driver
maybe in some experiment that comes close to the 43ms needed?

there is also the berry-fastloop option: https://tasmota.github.io/docs/Berry/#fast-loop you could try to use.

i dont know if a chain of timers (callback) can make a loop around 50ms, but i maybe  worth a try.
maybe you can make use of my experimening-code? https://github.com/arendst/Tasmota/discussions/18116#discussioncomment-5207047

**Reply by sfromis** [reply]
Regular timer callbacks are on the 50 ms ticks (meaning no faster than that)

**Reply by ozett** [reply]
![image](https://user-images.githubusercontent.com/5937811/224380496-d3157f7c-4f33-425e-83ee-3f3aa58b0106.png)
[link](https://community.home-assistant.io/t/people-counter-passing-through-a-door-using-vl53l0x-or-vl53l1x/312897)


it seems that the experimental setup will recognize people when passing through the two light barriers at a [maximum walking speed](https://github.com/arendst/Tasmota/discussions/18082#discussioncomment-5213428) of 5km/h  (==1.38m/s ==6cm/23ms) 
correct? 

>edit: calculated again roughly with max walking speed of 4km/h == 1m/s  ==   10cm/10ms
if this is better calculated, that would mean that:
if both photocells within a distance of 10 cm are blocked at the same time , that is detecting a person [pict. 3]. 
i think that means both blockings must be detected within a sensor-reading timeframe of 10ms (or less). 
difficult to do with berry/tasmota?


only at full running speed, after triggering photocell 1, about 23ms seconds later the photocell 2 must trigger. (and both signal of both tiggers must arrive a tasmota.  for that the triggers have to be in a distance of 6cm.  (and the person not thinner than 6cm?

at lower running speeds, between photocell 1 and photocell 2 more time would pass...
or more time would pass, if the cells have a bigger distance between each other
i.e. 46ms at a doubled distance of 12cm.

it the looptime of berry-driver is roughly 50ms and the distance between to triggers is roughly 12cm it could work...
it could also work with more time or shorter distance if people walk slower than 5km/h.. or calculations all wrong ?

**Reply by ozett** [reply]
after a sleepover i would also reconsider the actual structure for people counting.

* the barrier must not only be punctual or selective, 
it should not hang too low or too high, at best it should be a longer gate.
there are already such things on aliexpress. 

| hardware 1 | hardware 2|
| :---: | :---: |
<img src="https://user-images.githubusercontent.com/5937811/224479949-ef66876e-7090-4e46-946a-bb5e2edb6966.png" width="50%" height="50%"> | <img src="https://user-images.githubusercontent.com/5937811/224481089-c17bf4a6-c66b-409d-8878-5cd87d93d4f3.png" width="50%" height="50%">

* there are the borderline cases for detection with single/double beams: 
what about people who turn around, whose arms shake, or who simply stand still in the barrier?
or run fast as a bullet? [fast](https://td0g.ca/2020/04/19/ballistic-chronograph-mk2-diy/)   ->[lightgate](https://www.tindie.com/products/td0g/light-gate-sensor-for-ballistic-chronograph/)

* monitoring from the ceiling with thermo-sensor is a good idea. only needs one device. [here is such a solution](https://www.eledia.org/showcase/emuseum/feature-people-counter-infrared-scanning.xhtml):

<kbd align="center"> <img src="https://user-images.githubusercontent.com/5937811/224480121-1663d0cd-059b-4d58-8a0d-4488ecdca21d.png" width="50%" height="50%">
https://www.eledia.org/showcase/emuseum/feature-people-counter-infrared-scanning.xhtml

| single hardware | setup |
| :---: | :---: |
<img align="center" src="https://user-images.githubusercontent.com/5937811/224480452-ed759855-06ca-4eb1-8264-a83003444688.png" width=50% height=50%> | <img align="center" src="https://user-images.githubusercontent.com/5937811/224480682-5c8dbafa-a03b-4fa8-ab97-3859f7731644.png" > | 
D6T8L09 Thermischer MEMS-Präsenzsensor 1x8 Elemente, 5...50 °C | D6T1A02 Thermischer MEMS-Präsenzsensor 1x1 Element, -40...80 °C | 
https://docs.rs-online.com/e665/0900766b816cbfb2.pdf

now  it looks like the loop and tasmota-timing is the least problem in the whole hardware setup.

**Reply by ozett** [reply]
the person-counter idea can be realized by other sensors, wich are very,very interesting.
(a little bit off-topic for the original issue), but maybe worth listing here.

|single sensor above a door  |single sensor, better detection|
|:--:|:--:|
|https://www.youtube.com/watch?v=c91Ve-g0J2U  | **VL53L1X** gestures==directions with one sensor |
| <kbd> ![image](https://user-images.githubusercontent.com/5937811/224542172-1059fd91-bea1-4d69-abf7-5947e563c635.png) | ![image](https://user-images.githubusercontent.com/5937811/224542423-a41dd2df-3681-4bbe-9022-e0c341a24063.png) |
|-| https://github.com/stm32duino/VL53L1X |
|-| https://youtu.be/v8BYS-LokaE?t=305|
|-| https://github.com/Andrea-Fox/peopleCounter|
